tox.ini

```diff
@@ -80,3 +80,18 @@ commands =
     python -m setup checkdocs check --metadata --restructuredtext --strict --verbose
     python -m twine check .tox/dist/*
     python -m setuptools_scm ls
+
+[testenv:release]
+skip_install = True
+deps =
+    pep517>=0.5
+    twine>=1.13
+    path.py
+passenv =
+    TWINE_PASSWORD
+setenv =
+    TWINE_USERNAME = {env:TWINE_USERNAME:__token__}
+commands =
+    python -c "import path; path.Path('dist').rmtree_p()"
+    python -m pep517.build .
```

Comment:
I usually prefer having separate build and publish envs. Then we'd do smth like `TOXENV=build-dists,setup-check,release`.

Reply:
Three steps for one artifact? I dunno. That sounds overly complicated. Then you'd have to ensure that the artifacts from each invocation are shared across the three steps... and a manual release would then require the same steps (or at least the build/release steps). I can imagine a few scenarios where it might be useful to have separate steps, but 99% of the use-cases are going to be to invoke the suite of commands, so I'd prefer to keep it simple. And in that 1% of the cases where it's necessary to separate the concerns, one could use `tox --notest` and run the commands incrementally. Why not just keep it simple and concise?

Reply:
I think it does too much. I often want a command that gives me dists what I can then inspect with other tools. How about something like this https://github.com/pypa/twine/blob/5cb1867/tox.ini#L53-L61 then?

Reply:
Also, ideally in the future, I'd like to try out a better testing strategy: first build the artifacts, then pip-install from there and completely avoid lib imports from the repo.

Reply:
> I'd like to try out a better testing strategy

Yes, maybe. I'm heavily invested in the test-at-source methodology because it's faster. It's one of the reasons I prefer Python to (your language here). I do acknowledge there's weakness in testing against the source instead of the built artifact... but my instinct there would be to provide additional testing for the build artifact rather than encumber the development process with two sources of truth (source and installed code).

Reply:
Yeah, I think that for the development purposes it's more comfortable to rely on editable installs. And yes, I do agree that we need more artifact testing in CI than locally. I've been meaning to implement this for a while.
Here's a nice article that Paul published this topic last month: https://blog.ganssle.io/articles/2019/08/test-as-installed.html